// @arwes/core/lib/Blockquote/Blockquote.component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Blockquote = void 0;
/* @jsx jsx */
var react_1 = require("react");
var prop_types_1 = __importDefault(require("prop-types"));
var css_1 = require("@emotion/css");
var react_2 = require("@emotion/react");
var Animated_1 = require("../utils/Animated");
var appearTransitions_1 = require("../utils/appearTransitions");
var Blockquote_styles_1 = require("./Blockquote.styles");
var Blockquote = function (props) {
    var palette = props.palette, className = props.className, style = props.style, rootRef = props.rootRef, children = props.children;
    var theme = react_2.useTheme();
    var styles = react_1.useMemo(function () { return Blockquote_styles_1.generateStyles(theme, { palette: palette }); }, [theme, palette]);
    return (react_2.jsx("blockquote", { className: css_1.cx('arwes-blockquote', className), css: styles.root, style: style, ref: rootRef },
        react_2.jsx(Animated_1.Animated, { className: 'arwes-blockquote__bg', css: styles.bg, animated: {
                initialStyles: { opacity: 0 },
                entering: appearTransitions_1.transitionAppear,
                exiting: appearTransitions_1.transitionDisappear
            } }),
        react_2.jsx(Animated_1.Animated, { className: 'arwes-blockquote__line', css: styles.line, animated: {
                initialStyles: { opacity: 0, transform: 'scaleY(0)' },
                entering: [appearTransitions_1.transitionAppear, { scaleY: 1 }],
                exiting: [appearTransitions_1.transitionDisappear, { scaleY: 0 }]
            } }),
        react_2.jsx("div", { className: 'arwes-blockquote__content', css: styles.content }, children)));
};
exports.Blockquote = Blockquote;
Blockquote.propTypes = {
    palette: prop_types_1.default.string,
    className: prop_types_1.default.string,
    style: prop_types_1.default.object,
    rootRef: prop_types_1.default.any
};
