// @arwes/core/lib/Card/Card.component.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.Card = void 0;
/* @jsx jsx */
var react_1 = require("react");
var prop_types_1 = __importDefault(require("prop-types"));
var css_1 = require("@emotion/css");
var react_2 = require("@emotion/react");
var sounds_1 = require("@arwes/sounds");
var Text_1 = require("../Text");
var Card_styles_1 = require("./Card.styles");
var Card = function (props) {
    var animator = props.animator, image = props.image, title = props.title, options = props.options, landscape = props.landscape, hover = props.hover, rootRef = props.rootRef, className = props.className, style = props.style, children = props.children;
    var animate = animator.animate;
    var theme = react_2.useTheme();
    var bleeps = sounds_1.useBleeps();
    var styles = react_1.useMemo(function () { return Card_styles_1.generateStyles(theme, { animate: animate, landscape: landscape, hover: hover }); }, [theme, animate, landscape, hover]);
    var containerRef = react_1.useRef(null);
    animator.setupAnimateRefs(containerRef, theme, bleeps);
    return (react_2.jsx("article", { className: css_1.cx('arwes-card', className), css: styles.root, style: style, ref: rootRef },
        react_2.jsx("div", { className: 'arwes-card__container', css: styles.container, ref: containerRef },
            !!image && (react_2.jsx("div", { className: 'arwes-card__picture', css: styles.picture },
                react_2.jsx("img", { className: 'arwes-card__image', css: styles.image, style: { backgroundImage: "url(" + image.src + ")" }, src: image.src, alt: image.alt || '' }),
                react_2.jsx("div", { className: 'arwes-card__line arwes-card__line-picture', css: [styles.line, styles.linePicture] }))),
            react_2.jsx("div", { className: 'arwes-card__content', css: styles.content },
                react_2.jsx("div", { className: 'arwes-card__content-bg', css: styles.contentBg }),
                !!title && (react_2.jsx("header", { className: 'arwes-card__header', css: styles.header },
                    react_2.jsx(Text_1.Text, { as: 'h1', className: 'arwes-card__title', css: styles.title }, title))),
                react_2.jsx("div", { className: 'arwes-card__children', css: styles.children }, children),
                !!options && (react_2.jsx("div", { className: 'arwes-card__options', css: styles.options }, options)),
                react_2.jsx("div", { className: 'arwes-card__line arwes-card__line-content', css: [styles.line, styles.lineContent] })))));
};
exports.Card = Card;
Card.propTypes = {
    image: prop_types_1.default.shape({
        src: prop_types_1.default.string.isRequired,
        alt: prop_types_1.default.string
    }),
    title: prop_types_1.default.node,
    options: prop_types_1.default.node,
    landscape: prop_types_1.default.bool,
    hover: prop_types_1.default.bool,
    className: prop_types_1.default.string,
    style: prop_types_1.default.object,
    rootRef: prop_types_1.default.any
};
