// @arwes/core/lib/CodeBlock/CodeBlock.component.js
"use strict";
var __assign = (this && this.__assign) || function () {
    __assign = Object.assign || function(t) {
        for (var s, i = 1, n = arguments.length; i < n; i++) {
            s = arguments[i];
            for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p))
                t[p] = s[p];
        }
        return t;
    };
    return __assign.apply(this, arguments);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBlock = void 0;
/* @jsx jsx */
var react_1 = require("react");
var prop_types_1 = __importDefault(require("prop-types"));
var css_1 = require("@emotion/css");
var react_2 = require("@emotion/react");
var sounds_1 = require("@arwes/sounds");
var Text_1 = require("../Text");
var CodeBlock_styles_1 = require("./CodeBlock.styles");
var CodeBlock = function (props) {
    var animator = props.animator, lang = props.lang, contentTextProps = props.contentTextProps, children = props.children, className = props.className, style = props.style, rootRef = props.rootRef;
    var animate = animator.animate;
    var theme = react_2.useTheme();
    var bleeps = sounds_1.useBleeps();
    var styles = react_1.useMemo(function () { return CodeBlock_styles_1.generateStyles(theme, { animate: animate }); }, [theme, animate]);
    var containerRef = react_1.useRef(null);
    animator.setupAnimateRefs(containerRef, bleeps);
    return (react_2.jsx("div", { className: css_1.cx('arwes-code-block', className), css: styles.root, style: style, ref: rootRef },
        react_2.jsx("div", { className: 'arwes-code-block__container', css: styles.container, ref: containerRef },
            react_2.jsx("div", { className: 'arwes-code-block__bg', css: styles.bg }),
            react_2.jsx("div", { className: 'arwes-code-block__wrap', css: styles.wrap },
                react_2.jsx(Text_1.Text, __assign({}, contentTextProps, { className: 'arwes-code-block__content', css: styles.content }), children)),
            !!lang && (react_2.jsx("div", { className: 'arwes-code-block__lang', css: styles.lang },
                react_2.jsx("div", { className: 'arwes-code-block__lang-bg', css: styles.langBg }),
                react_2.jsx("div", { className: 'arwes-code-block__line arwes-code-block__line-lang', css: [styles.line, styles.lineLang] }),
                react_2.jsx(Text_1.Text, { blink: false }, lang))),
            react_2.jsx("div", { className: 'arwes-code-block__line arwes-code-block__line-top', css: [styles.line, styles.lineTop] }),
            react_2.jsx("div", { className: 'arwes-code-block__line arwes-code-block__line-bottom', css: [styles.line, styles.lineBottom] }))));
};
exports.CodeBlock = CodeBlock;
CodeBlock.propTypes = {
    lang: prop_types_1.default.string,
    contentTextProps: prop_types_1.default.object,
    className: prop_types_1.default.string,
    style: prop_types_1.default.object,
    rootRef: prop_types_1.default.any
};
CodeBlock.defaultProps = {
    contentTextProps: {
        as: 'pre',
        blink: false
    }
};
